src/vivarium/framework/results/context.py

```diff
@@ -202,52 +206,88 @@ def gather_results(
+            # Add new columns of mapped values to the population to prevent name collisions
+            new_column = get_mapped_col_name(stratification.name)
+            if new_column in population.columns:
+                raise ValueError(
+                    f"Stratification column '{new_column}' "
+                    "already exists in the state table or as a pipeline which is a required "
+                    "name for stratifying results - choose a different name."
+                )
+            population[new_column] = stratification(population)
+
         # Optimization: We store all the producers by pop_filter and stratifications
         # so that we only have to apply them once each time we compute results.
-        for (pop_filter, stratifications), observations in self.observations[
+        for (pop_filter, stratification_names), observations in self.observations[
             lifecycle_phase
         ].items():
             # Results production can be simplified to
             # filter -> groupby -> aggregate in all situations we've seen.
-            filtered_pop = self._filter_population(population, pop_filter, stratifications)
+            filtered_pop = self._filter_population(
+                population, pop_filter, stratification_names
+            )
             if filtered_pop.empty:
                 yield None, None, None
             else:
-                if stratifications is None:
+                if stratification_names is None:
                     pop = filtered_pop
                 else:
-                    pop = self._get_groups(stratifications, filtered_pop)
+                    pop = self._get_groups(stratification_names, filtered_pop)
                 for observation in observations:
-                    yield observation.observe(
-                        event, pop, stratifications
-                    ), observation.name, observation.results_updater
+                    results = observation.observe(event, pop, stratification_names)
+                    if results is not None:
+                        self._rename_stratification_columns(results)
+
+                    yield (results, observation.name, observation.results_updater)
 
-    @staticmethod
     def _filter_population(
-        population: pd.DataFrame, pop_filter: str, stratifications: tuple[str, ...]
+        self,
+        population: pd.DataFrame,
+        pop_filter: str,
+        stratifications: Optional[tuple[str, ...]],
     ) -> pd.DataFrame:
         """Filter the population based on the filter string as well as any
         excluded stratification categories
         """
         pop = population.query(pop_filter) if pop_filter else population.copy()
-        # Drop all rows in the 'stratifications' columns that have NaN values
         if stratifications:
-            pop = pop.dropna(subset=list(stratifications))
+            # Drop all rows in the mapped_stratification columns that have NaN values
```

Comment:
Nit: maybe explain that we are dropping these rows because being NaN means that the value was previously one of the excluded categories.